From 2cede636d8c795b3f458b90ab45f3d0ad5b2f258 Mon Sep 17 00:00:00 2001
From: milaq <mail@milaq.net>
Date: Sat, 21 Jan 2023 13:00:45 +0100
Subject: [PATCH] Add more hotkeys

* TAB/Shift+Tab: Up/down navigation
* Shift+Insert: Paste from primary clipboard
---
 dmenu.c | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/dmenu.c b/dmenu.c
index 5e694ec..d5863af 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -410,6 +410,16 @@ keypress(XKeyEvent *ev)
 		default:
 			return;
 		}
+	} else if (ev->state & ShiftMask) {
+		switch(ksym) {
+		case XK_ISO_Left_Tab:
+			ksym = XK_Up;
+			break;
+		case XK_Insert:
+			XConvertSelection(dpy, (ev->state & ShiftMask) ? clip : XA_PRIMARY,
+			                  utf8, utf8, win, CurrentTime);
+			return;
+		}
 	}
 
 	switch(ksym) {
@@ -509,19 +519,12 @@ insert:
 		/* fallthrough */
 	case XK_Down:
 	case XK_KP_Down:
+	case XK_Tab:
 		if (sel && sel->right && (sel = sel->right) == next) {
 			curr = next;
 			calcoffsets();
 		}
 		break;
-	case XK_Tab:
-		if (!sel)
-			return;
-		cursor = strnlen(sel->text, sizeof text - 1);
-		memcpy(text, sel->text, cursor);
-		text[cursor] = '\0';
-		match();
-		break;
 	}
 
 draw:
-- 
2.39.1

